src/Features/Core/Portable/Completion/CompletionItem.cs

```diff
@@ -89,7 +92,51 @@ public sealed class CompletionItem : IComparable<CompletionItem>
         /// <summary>
         /// Additional information attached to a completion item by it creator.
         /// </summary>
-        public ImmutableDictionary<string, string> Properties { get; }
+        public ImmutableDictionary<string, string> Properties
+        {
+            get
+            {
+                if (_propertiesAsImmutableDictionary is null)
+                    _propertiesAsImmutableDictionary = _properties.ToImmutableDictionary();
+
+                return _propertiesAsImmutableDictionary;
+            }
+        }
+
+        internal ImmutableArray<KeyValuePair<string, string>> GetProperties()
+        {
+            return _properties;
+        }
+
+        internal bool TryGetProperty(string name, [NotNullWhen(true)] out string? value)
+        {
+            if (_propertiesAsImmutableDictionary is not null)
+                return _propertiesAsImmutableDictionary.TryGetValue(name, out value);
+
+            foreach ((var propName, var propValue) in _properties)
+            {
+                if (name == propName)
+                {
+                    value = propValue;
+                    return true;
+                }
+            }
+
+            value = null;
+            return false;
+        }
+
+        internal string GetProperty(string name)
+        {
+            if (TryGetProperty(name, out var value))
+                return value;
+
+            // Let ImmutableDictionary handle throwing
+            if (_propertiesAsImmutableDictionary is not null)
+                return _propertiesAsImmutableDictionary[name];
+
+            throw new KeyNotFoundException($"Property {name} not found");
```

Comment:
Not sure what's the benefit of this either. I'd prefer to just throw, given it's the path we'd hit most of the time